File: 2005/Windows SDK NetFx 3.0/Technologies/DirectoryServices/Ldap_Exceptions/CS/Ldap_Exceptions/Ldap_Exceptions.cs
/*=====================================================================
  File:      Exceptions.cs

  Summary:   Demonstrates exceptions like LdapException, OperationException etc.

---------------------------------------------------------------------
  This file is part of the Microsoft .NET SDK Code Samples.

  Copyright (C) Microsoft Corporation.  All rights reserved.

This source code is intended only as a supplement to Microsoft
Development Tools and/or on-line documentation.  See these other
materials for detailed information regarding Microsoft code samples.

THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
PARTICULAR PURPOSE.
=====================================================================*/

// To compile: csc.exe /out:App.exe /r:System.DirectoryServices.Protocols.dll
//                                                          Exceptions.cs
//
// To run: App.exe <ldapServer> <user> <pwd> <domain>
//     Eg: App.exe myDC1.testDom.fabrikam.com  user1  secret@~1 testDom

using System;
using System.Net;
using System.Security.Permissions;
using SDS = System.DirectoryServices;


using System.DirectoryServices.Protocols;

[assembly: System.Reflection.AssemblyVersion("1.0.0.0")]
[assembly: SecurityPermission(SecurityAction.RequestMinimum, Execution = true)]
[assembly: SDS.DirectoryServicesPermission(SecurityAction.RequestMinimum)]


namespace Microsoft.Samples.DirectoryServices
{
    
public class App
{
    // static variables used throughout the sample code
    static LdapConnection ldapConnection;
    static string  ldapServer;
    static NetworkCredential credential;

    public static void Main(string[] args)
    {
        try
        {
            if(!GetParameters(args))
            {
                return;
            }
            
            CreateConnection();
            HandleLdapException();
            HandleDirectoryOperationException();
            HandlePlatformNotSupportedException();
            HandleBerConversionException();
            HandleArgumentNullException();
            HandleTlsOperationException();
            HandleInvalidOperationException();
            
            Console.WriteLine("\r\nApplication Finished Successfully!!!");
        }
        catch(Exception e)
        {
            Console.WriteLine("\r\nUnexpected exception occured:\r\n\t" + 
                              e.GetType().Name + ":" + e.Message);
        }        
    }



    static bool GetParameters(string[] args)
    {    
        if(args.Length != 4)
        {
            Console.WriteLine("Usage: App.exe <ldapServer> <user> " +
                                             "<pwd> <domain>");

            // return error 
            return false;
        }

        // initialize variables
        ldapServer = args[0];
        credential = new NetworkCredential(args[1], args[2], args[3]);

        // return success
        return true;
    }



    static void CreateConnection()
    {
        ldapConnection = new LdapConnection(ldapServer);        
        ldapConnection.Credential = credential;
        Console.WriteLine("LdapConnection is created successfully.");
    }

    static void HandleLdapException()
    {
        try
        {
            Console.WriteLine("\r\nTrying to connect to an unknown"+
                              " Ldap server...");

            LdapConnection conn = new LdapConnection("badServer");
            conn.Bind();
        }
        catch(LdapException e)
        {
            Console.WriteLine(e.GetType().Name + 
                              ": ErrorCode:"+ e.ErrorCode + 
                              ", Message:"  + e.Message);
        }


        try
        {
            Console.WriteLine("\r\nTrying to authenticate "+
                              "with bogus credentials...");
            
            ldapConnection.Credential = new NetworkCredential("bogusUser", 
                                                              "bogusPwd",
                                                              "bogusDom");
            ldapConnection.Bind();            
        }
        catch(LdapException e)
        {
            Console.WriteLine(e.GetType().Name + 
                              ": ErrorCode:"+ e.ErrorCode + 
                              ", Message:"  + e.Message);

            // revert back to original credentials supplied
            ldapConnection.Credential  = credential;
        }        
    }


    static void HandleDirectoryOperationException()
    {
        try
        {
            Console.WriteLine("\r\nSending a delete request "+
                                                    "with an invalid dn... ");
            
            ldapConnection.SendRequest(new DeleteRequest("#thisIsAnInValidDN#"));  
        }
        catch(DirectoryOperationException e)
        {
            DeleteResponse response = (DeleteResponse)e.Response;
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message);
            Console.WriteLine("Response: ResultCode:" + response.ResultCode +
                                    " : ErrorMessage:" + response.ErrorMessage);
        }
    }


    static void HandlePlatformNotSupportedException()
    {
        try
        {
            LdapConnection conn = new LdapConnection("someServer");
        }
        catch(PlatformNotSupportedException e)
        {
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message);

            // The application should handle this exception and 
            // fail gracefully on an unsupported platform
            Console.WriteLine("The application is not supported "+
                                       "on this platform and will shutdown...");
        }
    }
    

    static void HandleBerConversionException()
    {
        try
        {
            Console.WriteLine("\r\nTrying to decode a binary "+
                                    "value with an incorrect decode string...");

            BerConverter.Decode("{iii}",
                                new byte[]{0x00,0xFF,0x30,0x84,0x00,0x00,0x00});
        }
        catch(BerConversionException e)
        {
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message);
        }
    }


    static void HandleArgumentNullException()
    {
        try
        {
            Console.WriteLine("\r\nTrying to create "+
                                    "DirectoryAttribute with null values...");
            
            DirectoryAttribute attribute = new DirectoryAttribute(
                                                    "attrName", 
                                                    new string[]{"val1", 
                                                                 null, 
                                                                 "val2"});
        }
        catch(ArgumentNullException e)
        {
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message);
        }       
    }

    
    static void HandleTlsOperationException()
    {    
        try
        {
            Console.WriteLine("\r\nCalling StopTransportLayerSecurity()... ");
            ldapConnection.SessionOptions.StopTransportLayerSecurity();
        }
        catch(TlsOperationException e)
        {
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message);
        }
    }


    static void HandleInvalidOperationException()
    {
        try
        {
            Console.WriteLine("\r\nTrying anonymous authentication "+
                                                "with a non-null credential ");
            ldapConnection.AuthType = AuthType.Anonymous;
            ldapConnection.Credential = new NetworkCredential("userName",
                                                              "pwd",
                                                              "domain");
            ldapConnection.Bind();
            
        }
        catch(InvalidOperationException e)
        {
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message);
            // revert back
            ldapConnection.AuthType = AuthType.Negotiate;
            ldapConnection.Credential = credential;
        }    
    }

}// end class App

}

File: 2005/Windows SDK NetFx 3.0/Technologies/DirectoryServices/Ldap_Exceptions/VB/Ldap_Exceptions/Ldap_Exceptions.vb
' ===================================
'   File:      Exceptions.vb
'  
'   Summary:   Demonstrates exceptions like LdapException, 
'              DirectoryOperationException etc.
'  
' ---------------------------------------------------------------------
'   This file is part of the Microsoft .NET SDK Code Samples.
'  
'   Copyright (C) Microsoft Corporation.  All rights reserved.
'  
' This source code is intended only as a supplement to Microsoft
' Development Tools and/or on-line documentation.  See these other
' materials for detailed information regarding Microsoft code samples.
'  
' THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
' KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
' IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
' PARTICULAR PURPOSE.
' ===================================


 
' To compile: vbc.exe /out:App.exe /r:System.DirectoryServices.Protocols.dll 
'                                                                 Exceptions.vb
'
' To run: App.exe <ldapServer> <user> <pwd> <domain>
'     Eg: App.exe myDC1.testDom.fabrikam.com  user1  secret@~1 testDom

Option Strict On
Option Explicit On

Imports System
Imports System.Net
Imports System.Security.Permissions
Imports SDS = System.DirectoryServices

Imports System.DirectoryServices.Protocols

<Assembly: System.Reflection.AssemblyVersion("1.0.0.0")>
<Assembly: SecurityPermission(SecurityAction.RequestMinimum,Execution := true)>
<Assembly: SDS.DirectoryServicesPermission(SecurityAction.RequestMinimum)>

Namespace Microsoft.Samples.DirectoryServices

 
Public NotInheritable Class App
    ' static variables used throughout the sample code
    Shared ldapConnection As LdapConnection 
    Shared ldapServer As String  
    Shared credential As NetworkCredential 

    Private Sub New()
    End Sub

    Public Shared  Sub Main(ByVal args() As String)        
        Try
            if Not GetParameters(args) Then
                Return
            End If

            CreateConnection()
            HandleLdapException()
            HandleDirectoryOperationException()
            HandlePlatformNotSupportedException()
            HandleBerConversionException()
            HandleDirectoryArgumentNullException()
            HandleTlsOperationException()
            HandleInvalidOperationException()
 
            Console.WriteLine(vbCrLf + "Application Finished Successfully!!!")
        Catch e As Exception
            Console.WriteLine(vbCrLf + "Unexpected exception occured:" + _
                           vbCrLf + vbTab + e.GetType().Name + ":" + e.Message)
        End Try
    End Sub
 
 
 
    Shared Function GetParameters(args() As String) As Boolean
        If args.Length <> 4 Then
            Console.WriteLine("Usage: App.exe <ldapServer> <user> <pwd> <domain>")
            Return false
        End If
 
        ' initialize variables
        ldapServer = args(0)
        credential = New NetworkCredential(args(1), args(2), args(3))

        Return true
    End Function
 
 
 
    Shared  Sub CreateConnection()
        ldapConnection = New LdapConnection(ldapServer)        
        ldapConnection.Credential = credential
        Console.WriteLine("LdapConnection is created successfully.")
    End Sub
 
    Shared  Sub HandleLdapException()
        Try
            Console.WriteLine(vbCrLf + "Trying to connect to an "+ _
                                                       "unknown Ldap server...")

            Dim conn As LdapConnection  = New LdapConnection("badServer")
            conn.Bind()
        Catch e As LdapException
            Console.WriteLine(e.GetType().Name + _
                              ": ErrorCode:"+ e.ErrorCode.ToString() + _
                              ", Message:" + e.Message)
        End Try
 
 
        Try
            Console.WriteLine(vbCrLf + "Trying to authenticate "+ _
                                                    "with bogus credentials...")
                                                    
            ldapConnection.Credential = New NetworkCredential("bogusUser", _
                                                              "bogusPwd", _
                                                              "bogusDom")
            ldapConnection.Bind()
        Catch e As LdapException
            Console.WriteLine(e.GetType().Name + _
                              ": ErrorCode:"+ e.ErrorCode.ToString() + _
                              ", Message:" + e.Message)
 
            ' revert back to original credentials supplied
            ldapConnection.Credential  = credential
        End Try
    End Sub
 
 
    Shared  Sub HandleDirectoryOperationException()
        Try
            Console.WriteLine(vbCrLf + "Sending a delete request " + _
                                                    "with an invalid dn... ")

            ldapConnection.SendRequest(New DeleteRequest("#thisIsAnInValidDN#"))
        Catch e As DirectoryOperationException
            Dim response As DeleteResponse  = CType(e.Response, DeleteResponse)
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message)
            Console.WriteLine("Response: ResultCode:" + _
                                    response.ResultCode.ToString() + _
                                    " : ErrorMessage:" + response.ErrorMessage)
        End Try
    End Sub
 
 
    Shared  Sub HandlePlatformNotSupportedException()
        Try
            Dim conn As LdapConnection  = New LdapConnection("someServer")
        Catch e As PlatformNotSupportedException
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message)
 
            ' The application should handle this exception and 
            ' fail gracefully on an unsupported platform
            Console.WriteLine("The application is not supported on " + _
                                           "this platform and will shutdown...")
        End Try
    End Sub
 
 
    Shared  Sub HandleBerConversionException()
        Try
            Console.WriteLine(vbCrLf + "Trying to decode a binary " + _
                                    "value with an incorrect decode string...")

            BerConverter.Decode("{iiiiiiiii}",New Byte(){&H01,&H02,&H03,&H04, _
                                                         &H05,&H05,&H05,&H05, _
                                                         &H05,&H05,&H05})
        Catch e As BerConversionException
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message)
        End Try
    End Sub
 
 
    Shared  Sub HandleDirectoryArgumentNullException()
        Try
            Console.WriteLine(vbCrLf + "Trying to create " + _
                                       "DirectoryAttribute with null values...")

            Dim attribute As DirectoryAttribute = New DirectoryAttribute( _
                                                    "attrName", _
                                                    New String(){"val1", _
                                                                 Nothing, _
                                                                 "val2"})

        Catch e As ArgumentNullException
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message)
        End Try
    End Sub
 
 
    Shared  Sub HandleTlsOperationException()    
        Try
            Console.WriteLine(vbCrLf + "Calling StopTransportLayerSecurity()...")
            ldapConnection.SessionOptions.StopTransportLayerSecurity()
        Catch e As TlsOperationException
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message)
        End Try
    End Sub
 
 
    Shared  Sub HandleInvalidOperationException()
        Try
            Console.WriteLine(vbCrLf + "Trying anonymous authentication " + _
                                                  "with a non-null credential ")

            ldapConnection.AuthType = AuthType.Anonymous
            ldapConnection.Credential = New NetworkCredential("userName", _
                                                              "pwd", _
                                                              "domain")
            ldapConnection.Bind()
 
        Catch e As InvalidOperationException
            Console.WriteLine(e.GetType().Name + ": Message:" + e.Message)
            ' revert back
            ldapConnection.AuthType = AuthType.Negotiate
            ldapConnection.Credential = credential
        End Try
    End Sub
 
End Class
' end class App

End Namespace
